helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -757,9 +753,20 @@ private void handleEvent(ClusterEvent event, BaseControllerDataProvider dataProv
     }
     event.addAttribute(AttributeName.ControllerDataProvider.name(), dataProvider);
 
-    logger.info(String.format("START: Invoking %s controller pipeline for cluster %s event: %s  %s",
-        manager.getClusterName(), dataProvider.getPipelineName(), event.getEventType(),
-        event.getEventId()));
+    // If manager session changes, no need to run pipeline for the stale event.
+    Optional<String> eventSessionId = event.getAttribute(AttributeName.EVENT_SESSION.name());
+    String managerSessionId = manager.getSessionId();
```

Comment:
If I understand correctly, the sessionId here can still be stale. eventually, we need sessionaware write of zkclient used by ZkHelixManager to ensure correctness, right?

Reply:
I think there are  other event not trigger from ZkClient, such as periodical timer triggered, or onDemandtimer triggered. Do we need special treatment here? or Special treatment when event to controller is constructed?

Reply:
@kaisun2000 Here if session id from manager is different from event session -- the session is changed. We check it and stop processing the event or sending messages.
The check in Zkclient is the last step. Most of the time this check here will stop the stale session sending messages.
Regarding other events, each time an event is created, we add the event_session to it. The event session is from cluster manager.

`if (!eventSessionId.isPresent() || !eventSessionId.get().equals(managerSessionId)) {` only it's leader and session doesn't change, it continues, otherwise, pipeline is not invoked.